worker/src/main/java/gov/cms/ab2d/worker/service/JobProcessingServiceImpl.java

```diff
@@ -30,28 +59,167 @@ public Job putJobInProgress(String jobId) {
+        final List<Sponsor> childrenSponsors = sponsorRepository.findByParent(sponsor);
+        return childrenSponsors.stream()
+                .map(s -> s.getAttestedContracts())
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
+    }
+
+
+    private List<JobOutput> processContract(final Path outputDir, Contract contract) {
+
+        var contractNumber = contract.getContractNumber();
+        var outputFile = fileService.createFile(outputDir, contractNumber + OUTPUT_FILE_SUFFIX);
+        var errorFile = fileService.createFile(outputDir, contractNumber + ERROR_FILE_SUFFIX);
+
+        var patientsByContract = beneficiaryAdapter.getPatientsByContract(contractNumber);
+        var patients = patientsByContract.getPatients();
+        int patientCount = patients.size();
+
+        // A mutex lock that all threads for a contract uses while writing into the shared files
+        var lock = new ReentrantLock();
+        var futureResourcesHandles = patients.stream()
+                .map(patient -> patientClaimsProcessor.process(patient.getPatientId(), lock, outputFile, errorFile))
+                .collect(Collectors.toList());
+
+        int errorCount = processHandles(futureResourcesHandles);
+        while (!futureResourcesHandles.isEmpty()) {
+            sleep();
+            errorCount += processHandles(futureResourcesHandles);
+        }
+
+        final List<JobOutput> jobOutputs = new ArrayList<>();
+        if (errorCount < patientCount) {
+            final JobOutput jobOutput = createPartialJobOutput(outputFile);
+            jobOutput.setError(false);
+            jobOutputs.add(jobOutput);
+        }
+        if (errorCount > 0) {
+            final JobOutput jobOutput = createPartialJobOutput(errorFile);
+            jobOutput.setError(true);
+            jobOutputs.add(jobOutput);
+        }
+
+        return jobOutputs;
+    }
+
+    private int processHandles(List<Future<Integer>> futureResourcesHandles) {
+        int errorCount = 0;
+
+        var iterator = futureResourcesHandles.iterator();
+        while (iterator.hasNext()) {
+            var future = iterator.next();
+            if (future.isDone()) {
+                try {
+                    var responseCount = future.get();
+                    if (responseCount > 0) {
+                        errorCount += responseCount;
+                    }
+                } catch (InterruptedException  e) {
+                    ++errorCount;
+                    log.error("interrupted exception while processing patient ", e);
+                } catch (ExecutionException e) {
+                    ++errorCount;
+                    log.error("exception while processing patient ", e.getCause());
+                }
+                iterator.remove();
+            }
+        }
+
+        return errorCount;
+    }
+
+    private JobOutput createPartialJobOutput(Path outputFile) {
```

Comment:
Based on how this method is being used, the value of the `error` flag is know by the time this method is called. So you might as well just call it `createJobOutput` and pass `isError` as a parameter.

Reply:
done